cpp/include/cuvs/cluster/kmeans_mg.hpp

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+#include <cuvs/cluster/kmeans.hpp>
+#include <raft/core/device_mdspan.hpp>
+#include <raft/core/host_mdspan.hpp>
+#include <raft/core/resources.hpp>
+
+namespace cuvs::cluster::kmeans::mg {
```

Comment:
One of the nice benefits to using `raft::resources` and further using a PIMPL design for the `raft::comms` is that we shouldn't be coupled to NCCL at all in the API layer. For this reason, I think we should expose a single unified API for both the single-gpu and multi-gpu kmeans and just route to the proper implementation based on the presence of the `raft::comms` object in the `raft::resources` object. Does that make sense?

Reply:
The other benefit to this is that we still expose the same symbols from the API side and can consolidate everything down (since the MNMG version really just uses primitives from the SG version, we don't need different instantiations for those).

Reply:
I added this in the last commit https://github.com/rapidsai/cuvs/pull/348/commits/d78dc8c2b9281b685de014b787f51c0b26f83d7e - I like how this simplifies the public api a bunch here =)